client/core/core.go

```diff
@@ -7437,3 +7479,223 @@ func (c *Core) findActiveOrder(oid order.OrderID) (*trackedTrade, error) {
+			tOldest = t
+		}
+	}
+	tilNext := fiatRateRequestInterval - time.Since(tOldest)
+	if tilNext < minTick {
+		tilNext = minTick
+	}
+	return time.NewTimer(tilNext)
+}
+
+// FiatRateSources returns a list of fiat rate sources and their individual
+// status.
+func (c *Core) FiatRateSources() map[string]bool {
+	c.ratesMtx.RLock()
+	defer c.ratesMtx.RUnlock()
+	rateSources := make(map[string]bool, len(fiatRateFetchers))
+	for token := range fiatRateFetchers {
+		rateSources[token] = c.fiatRateSources[token] != nil
+	}
+	return rateSources
+}
+
+// fiatConversions returns fiat rate for all supported assets that have a
+// wallet. It also checks if fiat rates are expired and does some clean-up.
+func (c *Core) fiatConversions() map[uint32]float64 {
+	supportedAssets := c.SupportedAssets()
+
+	c.ratesMtx.RLock()
+	defer c.ratesMtx.RUnlock()
+	disabledCount := 0
+	fiatRatesMap := make(map[uint32]float64, len(supportedAssets))
+	for assetID := range supportedAssets {
+		var rateSum float64
+		var sources int32
+		for token, source := range c.fiatRateSources {
+			// Remove fiat rate source with expired exchange rate data.
+			if source.isExpired(fiatRateDataExpiry) {
+				delete(c.fiatRateSources, token)
+				disabledCount++
+				continue
+			}
+			rateInfo := source.assetRate(assetID)
+			if rateInfo != nil && time.Since(rateInfo.lastUpdate) < fiatRateDataExpiry {
+				sources++
+				rateSum += rateInfo.rate
+			}
+		}
+		if rateSum != 0 {
+			fiatRatesMap[assetID] = rateSum / float64(sources) // get average rate.
+		}
+	}
+
+	// Ensure disabled fiat rate fetchers are saved to database.
+	if disabledCount > 0 {
+		c.saveDisabledRateSources()
+	}
+
+	return fiatRatesMap
+}
+
+// ToggleRateSourceStatus toggles a fiat rate source status. If disable is true,
+// the fiat rate source is disabled, otherwise the rate source is enabled.
+func (c *Core) ToggleRateSourceStatus(source string, disable bool) error {
+	if disable {
+		return c.disableRateSource(source)
+	}
+	return c.enableRateSource(source)
+}
+
+// enableRateSource enables a fiat rate source.
+func (c *Core) enableRateSource(source string) error {
+	c.ratesMtx.Lock()
+	defer c.ratesMtx.Unlock()
+
+	// Check if it's an invalid rate source or it is already enabled.
+	rateFetcher, found := fiatRateFetchers[source]
+	if !found {
+		return errors.New("cannot enable unknown fiat rate source")
+	} else if c.fiatRateSources[source] != nil {
+		return nil // already enabled.
+	}
+
+	// Build fiat rate source.
+	c.fiatRateSources[source] = newCommonRateSource(rateFetcher)
+
+	// If this is our first fiat rate source, start fiat rate fetcher goroutine,
+	// else fetch rates.
+	if len(c.fiatRateSources) == 1 {
+		c.fetchFiatExchangeRates()
+	} else {
+		source := c.fiatRateSources[source]
+		supportedAssets := c.SupportedAssets()
+		go func(source *commonRateSource) {
+			ctx, cancel := context.WithTimeout(c.ctx, 4*time.Second)
+			defer cancel()
+			source.refreshRates(ctx, c.log, supportedAssets)
```

Comment:
On this path, it looks like there won't be a fiatRateUpdate note until the next regular tick.  Intended?

Reply:
> On this path, it looks like there won't be a fiatRateUpdate note until the next regular tick. Intended?

Yes, fetching rates for the newly enabled source will ensure the next tick includes the source rates when computing the average fiat rate for an asset.